packages/belos/src/BelosGmresPolySolMgr.hpp

```diff
@@ -90,39 +91,53 @@ class GmresPolySolMgrPolynomialFailure : public BelosError {public:
-/// effective as the equivalent number of GMRES iterations.
-///
-/// We call solvers that take this approach "seed solvers."  Belos
-/// implements both a Block GMRES seed solver (this class) and a
-/// CG-based seed solver (Belos::PCPGSolMgr).
+/// The GMRES polynomial solver manager can perform two types of linear
+/// solves. First the solver runs block GMRES, storing the resulting 
+/// coefficients (or roots), which can be used to form a matrix polynomial.
+/// It then can reuse this polynomial as either a surrogate operator, or
+/// as a preconditioner for an outer solver.
+/// By applying the GMRES polynomial as an operator or preconditioner, one 
+/// avoids the cost of the inner products and norms in GMRES, thus reducing
+/// communication costs.  
+//
+/// The GMRES polynomial can be created in conjunction with any standard preconditioner.
+/// Simply pass the preconditioner to the LinearProblem before calling the GmresPolySolMgr
+/// and your preconditioner will be combined with the polynomial automatically.
 ///
 /// Here is a list of all the parameters that this solver accepts:
-///   - "Convergence Tolerance" (\c MagnitudeType): The level that
+///   - "Polynomial Type" (\c std::string): The desired polynomial type: 
+///      Roots, Arnoldi, or Gmres.  Default: "Roots"
+///   - "Polynomial Tolerance" (\c MagnitudeType): The level that
 ///     residual norms must reach to decide convergence. Default:
 ///     1e-8.
-///   - "Block Size" (\c int): The block size to be used by the
-///     underlying block GMRES solver. Default: 1 (which means don't
-///     use blocks).
-///   - "Num Blocks" (\c int): The restart length; that is, the number
-///     of blocks allocated for the Krylov basis. Default: 300.
-///   - "Maximum Iterations" (\c int): The maximum number of
-///     iterations GMRES is allowed to perform, across all restarts.
-///     Default: 1000.
-///   - "Maximum Restarts" (\c int): The maximum number of restarts
-///     the underlying solver is allowed to perform.  This does
-///     <i>not</i> include the first restart cycle.  Default: 20.
+///   - "Maximum Degree" (\c int): Requested maximum degree for the polynomial. Default: 25
+///   - "Random RHS" (\c bool): to generate the polynomial using a random vector. Default: true
+///   - "Add Roots" (\c bool): to add roots to the polynomial as needed for stability. Default: true
+///   - "Damp Poly" (\c bool): to damp polynomial. Default: false
 ///   - "Orthogonalization" (\c std::string): The desired
-///     orthogonalization method.  Default: "DGKS".
+///     orthogonalization method to create polynomial.  Default: "DGKS".
 ///   - "Verbosity" (Belos::MsgType): A sum of Belos::MsgType values
 ///     specifying what kinds of messages to print.  Default:
 ///     Belos::Errors.
-///   - "Output Style" (Belos::OutputType): The output style.
-///     Default: Belos::General.
+///   - "Outer Solver" (\c std::string): Name of the outer solver in Belos solver factory.
```

Comment:
Regarding Risk (2): The default "outer solver" option is to use no outer solver and only apply the polynomial. 

Regarding Risk (3): I'm not certain this would always have to be handled as a special case.  One option (perhaps with slightly more overhead) is to run everything through the polynomial solver with default polynomial degree of zero (thus no polynomial preconditioning).  So, for example, you could use the polynomial solver manager with outer solver GMRES and pass in an ILU preconditioner to get the same ILU preconditioned GMRES as you would with the GMRES solver manager.  

As for adding the polynomial to Ifpack2: I talked briefly about this with @hkthorn.  While there are some good reasons to incorporate the polynomial into Ifpack2, current Ifpack2 capabilities do not allow us to combine the polynomial with other preconditioners as needed.  Thus, for current research and testing purposes, we plan to stick with the Belos implementation.

Reply:
Thanks for your answer!

@jennloe wrote:
> Regarding Risk (3): I'm not certain this would always have to be handled as a special case.

If I turn off the "outer solver" option, can I then use your polynomial preconditioner as an arbitrary preconditioner, with any Belos solver that I've constructed myself?  If so, then that gets rid of the concern about an input deck special case.

> ... current Ifpack2 capabilities do not allow us to combine the polynomial with other preconditioners as needed.

Subclasses of `Ifpack2::Details::NestedPreconditioner` can take an "inner preconditioner."  `Ifpack2::AdditiveSchwarz` is a subclass of NestedPreconditioner, and uses the "inner preconditioner" as its subdomain solver.  Would that provide the capability you seek?

In any case, I'm perfectly happy with your polynomial preconditioner staying in Belos.  It's not specific to Tpetra and doesn't need to depend on it.

Reply:
By default there is no "outer solver", so the Belos solver is the polynomial precondioner, which can be used as an arbitrary preconditioner with any other solver (or as a surrogate operator for inv(A), as it is in my thesis).